gpgwrapper/parser.go

```diff
@@ -0,0 +1,160 @@
+package gpgwrapper
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Parse the output of --with-colons --list-secret keys and return only valid
+// keys (not revoked, not expired) as []SecretKeyListing
+// For the format of the colon-delimited string, see:
+// https://github.com/gpg/gnupg/blob/master/doc/DETAILS
+
+func parseListSecretKeys(colonDelimitedString string) ([]SecretKeyListing, error) {
+	parser := listSecretKeysParser{}
+
+	for _, line := range strings.Split(colonDelimitedString, "\n") {
+		parser.PushLine(strings.Split(line, ":"))
+	}
+	parser.End()
+
+	return parser.Keys, nil
+}
+
+// listSecretKeysParser takes a line at a time from the colon-delimited output
+// format of gpg --list-secret-keys.
+// Because different parts of the key are on different lines (created,
+// fingerprint, each uid), this object builds up a partial key as lines are
+// pushed.
+// When a line signifying a new key is pushed, or when End is called, the
+// partial key is checked for validity and added to Keys.
+
+type listSecretKeysParser struct {
+	partialKey *SecretKeyListing
+	Keys       []SecretKeyListing
+}
+
+func (p *listSecretKeysParser) PushLine(cols []string) {
+
+	typeOfRecord := cols[0]
+
+	switch typeOfRecord {
+	case "sec":
+		p.handleSecretPrimaryKeyLine(cols)
+		return
+
+	case "fpr":
+		p.handleFingerprintLine(cols)
+		return
+
+	case "uid":
+		p.handleUidLine(cols)
+		return
+	}
+}
+
+func (p *listSecretKeysParser) End() {
+	p.addCurrentKeyToList()
+}
+
+func (p *listSecretKeysParser) handleSecretPrimaryKeyLine(cols []string) {
+	p.addCurrentKeyToList()
+
+	validity := cols[1]
+	if validity == "r" || validity == "e" || validity == "n" {
+		// Ignore this invalid primary key.
+		// https://github.com/gpg/gnupg/blob/master/doc/DETAILS#field-2---validity
+		return
+	}
+
+	createdTime, err := parseTimestamp(cols[5])
+
+	if err != nil {
+		panic(err) // TODO: improve
```

Comment:
Still TODO?

Reply:
Fixed & test added in https://github.com/fluidkeys/fluidkeys/pull/48/commits/ef59e203fad63eb821b4dd6bd981392fdac1566f